modules/widgets/contact-info.php

```diff
@@ -379,16 +349,85 @@ function urlencode_address( $address ) {
 			return urlencode( $address );
 		}
 
+		/**
+		 * Returns the instance's updated 'goodmap' value.
+		 *
+		 * @param array $old_instance Old configuration values.
+		 * @param array $instance Current configuration values.
+		 *
+		 * @return bool|string The instance's updated 'goodmap' value. The value is true if
+		 * $instance can display a good map. If not, returns an error message.
+		 */
+		function update_goodmap( $old_instance, $instance ) {
+			/*
+			 * If we have no address or don't want to show a map,
+			 * no need to check if the map is valid.
+			 */
+			if ( empty( $instance['address'] ) || 0 === $instance['showmap'] ) {
+				return false;
+			}
+
+			/*
+			 * If there have been any changes that may impact the map in the widget
+			 * (adding an address, address changes, new API key, API key change)
+			 * then we want to check whether our map can be displayed again.
+			 */
+			if (
+				! isset( $instance['goodmap'] )
+				|| ! isset( $old_instance['address'] )
+				|| $this->urlencode_address( $old_instance['address'] ) !== $this->urlencode_address( $instance['address'] )
+				|| ! isset( $old_instance['apikey'] )
+				|| $old_instance['apikey'] !== $instance['apikey']
+			) {
+				return $this->has_good_map( $instance );
+			} else {
+				return $instance['goodmap'];
+			}
+		}
+
 		/**
 		 * Check if the instance has a valid Map location.
 		 *
 		 * @param array $instance Widget instance configuration.
 		 *
-		 * @return bool Whether or not there is a valid map.
+		 * @return bool|string Whether or not there is a valid map. If not, return an error message.
 		 */
 		function has_good_map( $instance ) {
-			// The lat and lon of an address that could not be plotted will have values of 0 and 0.
-			return ! ( '0' == $instance['lat'] && '0' == $instance['lon'] );
+			/** This filter is documented in modules/widgets/contact-info.php */
+			$api_key = apply_filters( 'jetpack_google_maps_api_key', $instance['apikey'] );
+			if ( ! empty( $api_key ) ) {
+				$path     = add_query_arg(
+					array(
+						'q'   => rawurlencode( $instance['address'] ),
+						'key' => $api_key,
+					),
+					'https://www.google.com/maps/embed/v1/place'
+				);
+				$response = wp_remote_get( esc_url_raw( $path ) );
+
+				if ( 200 === wp_remote_retrieve_response_code( $response ) ) {
+					return true;
+				} else {
+					return wp_remote_retrieve_body( $response );
+				}
+			}
+
+			return __( 'Please enter a valid Google API Key.', 'jetpack' );
+		}
+
+		/**
+		 * Check the Google Maps API key after an Ajax call from the widget's admin form in
+		 * the Customizer preview.
+		 */
+		function ajax_check_api_key() {
+			if ( isset( $_POST['apikey'] ) ) {
```

Comment:
Coming back to this, I think we should create a nonce, use it with our requests, check against them with `check_ajax_referer`, and check for `current_user_can( 'customize' )` as well.

If we don't protect access to those requests, we'll run the risk of having non logged-in users making requests instead of the admin.

Reply:
Good point. I'll add these checks.